File: src/main/kotlin/philo/peanutbox/core/feature/classifier/DiTypeClassifier.java
package philo.peanutbox.core.feature.classifier;

import static java.util.Arrays.stream;

import java.lang.reflect.Constructor;
import nextstep.web.annotation.GiveMePeanut;

public abstract class DiTypeClassifier implements TypeClassifier {

    public static boolean isDefaultConstructorInjection(Class<?> peanutClass) {
        
        return hasDefaultConstructor(peanutClass) && !hasFieldInjectionAnnotation(peanutClass);
    }

    public static boolean isFieldInjection(Class<?> peanutClass) {
        
        return hasDefaultConstructor(peanutClass) && hasFieldInjectionAnnotation(peanutClass);
    }

    public static boolean isConstructorWithArgumentsInjection(Class<?> peanutClass) {

        return hasConstructorWithArguments(peanutClass) && !hasFieldInjectionAnnotation(peanutClass);
    }

    private static boolean hasDefaultConstructor(Class<?> type) {
        
        try {
            type.getDeclaredConstructor();
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    private static boolean hasFieldInjectionAnnotation(Class<?> type) {
        
        return stream(type.getDeclaredFields())
                .anyMatch(field -> field.isAnnotationPresent(GiveMePeanut.class));
    }

    private static boolean hasConstructorWithArguments(Class<?> peanutClass) {

        Constructor<?> constructor = peanutClass.getDeclaredConstructors()[0];
        Class<?>[] parameterClasses = constructor.getParameterTypes();

        return parameterClasses.length >= 1;
    }
}

File: src/main/kotlin/philo/peanutbox/core/feature/classifier/DiTypeClassifier.kt
package philo.peanutbox.core.feature.classifier

import philo.peanutbox.core.annotation.GiveMePeanut
import java.lang.reflect.Field
import java.util.*

object DiTypeClassifier : TypeClassifier {
    fun isDefaultConstructorInjection(peanutClass: Class<*>): Boolean {
        return hasDefaultConstructor(peanutClass) && !hasFieldInjectionAnnotation(peanutClass)
    }

    fun isFieldInjection(peanutClass: Class<*>): Boolean {
        return hasDefaultConstructor(peanutClass) && hasFieldInjectionAnnotation(peanutClass)
    }

    fun isConstructorWithArgumentsInjection(peanutClass: Class<*>): Boolean {
        return hasConstructorWithArguments(peanutClass) && !hasFieldInjectionAnnotation(peanutClass)
    }

    private fun hasDefaultConstructor(type: Class<*>): Boolean {
        return try {
            type.getDeclaredConstructor()
            true
        } catch (e: Exception) {
            false
        }
    }

    private fun hasFieldInjectionAnnotation(type: Class<*>): Boolean {
        return Arrays.stream<Field>(type.getDeclaredFields())
            .anyMatch { field: Field ->
                field.isAnnotationPresent(
                    GiveMePeanut::class.java
                )
            }
    }

    private fun hasConstructorWithArguments(peanutClass: Class<*>): Boolean {
        val constructor = peanutClass.getDeclaredConstructors()[0]
        val parameterClasses = constructor.parameterTypes
        return parameterClasses.size >= 1
    }
}
